flight/mcu/lib/FifoBuffer/FifoBuffer.h

```diff
@@ -1,67 +1,163 @@
-//
-// Created by 16182 on 9/27/2021.
-//
-
 #ifndef TARS_SOFTWARE_FIFO_H
 #define TARS_SOFTWARE_FIFO_H
 
-#include <cstdint>
-#include <cstring>
+#include <ChRt.h>
 
-#ifdef COMPILE_LOCAL
-#include <mutex>
-#endif
+template<typename T, size_t max_size>
+class FifoView;
 
-#ifdef COMPILE_TARGET
-#include <ChRt.h>
-#endif
-
-// only works with types that can be copied with memcpy
-class GenericFifoBuffer {
-   public:
-    // buff_size = size of array in elements
-    // data_size = size of array element in bytes
-    GenericFifoBuffer(void* arr, uint16_t buff_size, uint16_t data_size) {
-        capacity_ = buff_size;
-        cur_length_ = 0;
-        head_idx_ = 0;
-        tail_idx_ = 0;
-        data_size_ = data_size;
-        arr_ = arr;
+template<typename T, size_t max_size>
+class FifoBuffer {
+public:
+    friend class FifoView<T, max_size>;
+
+    MUTEX_DECL(lock);
+
+    FifoBuffer() = default;
+
+    bool push(T const& element) {
+        chMtxLock(&lock);
+        arr[tail_idx++] = element;
+        if (tail_idx == max_size) {
+            tail_idx = 0;
+        }
+        count++;
+        chMtxUnlock(&lock);
+        return true;
+    }
+
+    bool read(T& read_to) {
+        chMtxLock(&lock);
+        if (count < 1) {
+            return false;
+        }
+        read_to = arr[tail_idx];
+        chMtxUnlock(&lock);
+        return true;
+    }
+
+    /**
+     * @brief Reads a range of items into a passed array, which can be larger than the actual count of items.
+     *
+     * @param write_to An array of at least n items to write to
+     * @param start The index to start reading from (inclusive)
+     * @param end The index to stop reading at (exclusive)
+     * @return How many items were actually read
+     */
+    size_t readSlice(T* write_to[], size_t start, size_t end) {
+        chMtxLock(&lock);
+        size_t i = 0;
+        if (count < max_size) {
+            size_t idx = start;
+            while (idx != tail_idx && idx < end) {
+                write_to[i++] = &arr[idx++];
+            }
+        } else {
+            // if we are already wrapping, head == tail
+            size_t idx = tail_idx + start;
+            if (idx >= max_size) {
+                idx -= max_size;
+            }
+            size_t end_idx = tail_idx + end;
+            if (end_idx >= max_size) {
+                end_idx -= max_size;
+            }
+            while (idx != tail_idx && idx < end_idx) {
+                write_to[i++] = &arr[idx++];
+            }
+        }
+        chMtxUnlock(&lock);
+        return i;
```

Comment:
This function seems flawed to me, why is it writing pointers to each element out to the array instead of values. This introduces thread safety issues where the pointers point to elements that can be mutated by other threads. The interface seems wrong as well because it is a circular buffer, but readSlice only reads a linear slice.

Reply:
Could you take a look at this as well, I've changed it

Reply:
I don't think its quite right. Returning pointers to elements instead of elements by value creates threading issues. Using start + end indices to indicated ranges doesn't make sense either given that its a circular buffer. I think start + length makes more sense here. You also can't assume that headidx = tailidx-1, that would only be true if the buffer was all the way full.

Reply:
I think the issue comes from using fifobuffer to do two different things, data log uses it as a queue of items to log to the sd card while gnc uses it as a buffer of the most recent sensor readings. I don't think that these two functions necessarily work well together as part of the same class. I can remove the changes to fifo buffer from this pr and make a separate pr to look at just those changes so this pr focuses on just removing the pointer struct.

Reply:
Eh, I think I have this almost correct. We'll need something like readSlice anyways, since the HistoryBufferFSM depends on having a buffer of past readings.  

I've changed the logic again.